livekit-android-sdk/src/main/java/io/livekit/android/room/RTCEngine.kt

```diff
@@ -252,60 +256,84 @@ internal constructor(
      */
     internal fun reconnect() {
+        if (reconnectingJob != null) {
+            return
+        }
+        if (this.isClosed) {
+            return
+        }
         val url = sessionUrl
         val token = sessionToken
         if (url == null || token == null) {
             LKLog.w { "couldn't reconnect, no url or no token" }
             return
         }
-        if (iceState == IceState.DISCONNECTED || wsRetries >= MAX_SIGNAL_RETRIES) {
-            LKLog.w { "could not connect to signal after max attempts, giving up" }
-            close()
-            listener?.onDisconnect("could not reconnect after limit")
-            return
-        }
 
-        var startDelay = wsRetries.toLong() * wsRetries * 500
-        if (startDelay > 5000) {
-            startDelay = 5000
-        }
-        coroutineScope.launch {
-            delay(startDelay)
-            if (iceState == IceState.DISCONNECTED) {
-                LKLog.e { "Ice is disconnected" }
-                return@launch
-            }
+        val job = coroutineScope.launch {
+            listener?.onReconnecting()
+
+            for (wsRetries in 0 until MAX_SIGNAL_RETRIES) {
+                var startDelay = wsRetries.toLong() * wsRetries * 500
+                if (startDelay > 5000) {
+                    startDelay = 5000
+                }
+
+                LKLog.i { "Reconnecting to signal, attempt ${wsRetries + 1}" }
+
+                delay(startDelay)
+                try {
+                    client.reconnect(url, token)
+                } catch (e: Exception) {
+                    // ws reconnect failed, retry.
+                    continue
+                }
+
+                LKLog.v { "ws reconnected, restarting ICE" }
+                listener?.onSignalConnected()
+
+                subscriber.prepareForIceRestart()
+                iceState = IceState.RECONNECTING
+                // trigger publisher reconnect
+                // only restart publisher if it's needed
+                if (hasPublished) {
+                    negotiate()
+                }
 
-            client.reconnect(url, token)
-
-            LKLog.v { "reconnected, restarting ICE" }
-            wsRetries = 0
-
-            // trigger publisher reconnect
-            subscriber.prepareForIceRestart()
-            // only restart publisher if it's needed
-            if (hasPublished) {
-                publisher.negotiate(
-                    getPublisherOfferConstraints().apply {
-                        with(mandatory) {
-                            add(
-                                MediaConstraints.KeyValuePair(
-                                    MediaConstraintKeys.ICE_RESTART,
-                                    MediaConstraintKeys.TRUE
-                                )
-                            )
-                        }
+                // wait until ICE connected
+                val endTime = SystemClock.elapsedRealtime() + MAX_ICE_CONNECT_TIMEOUT_MS;
```

Comment:
I just realised yesterday that Safari is quite slow to connect. With a 500 ms latency up & down in network link conditioner, the JS SDK default time out of 5 seconds was not enough for Safari (it was taking close to 9). Not sure what this constant is in mobile. But, keeping a larger number like 15/20 seconds would be good to cover long delay networks.

Reply:
500ms up & down means more than 1s rtt, and ice & dtls handshake may take at least 6*rtt I think. If some packets is lost, things became worse. 15/20 seconds make sense for this condition, but I doubt whether user can wait this long time for connecting(he don't know what is going, just wait ...)